### PR TITLE
SNO-144-redirect-downloads-through-nginx-or-apache

### PR DESCRIPTION
**Pull request by lrowe**
Instead of hard coding X-Accel-Redirect make it configurable so Apache mod_wsgi Location header can be used.
When unset proxy through Python process.

### CODE REVIEW

**Comment by lrowe** [comment]
Note: will require adding the setting to ini files in encoded.

**Comment by hitz** [comment]
@lrowe This looks pretty straightforward.  Is there anything to test?   I assume @cricketsloan still wants it?

**Comment by lrowe** [comment]
I think this is generally useful as it makes it possible to host only in Apache. Do make sure that the configuration change makes it through to encoded too as without it the file is proxied through Python.

To test curl a localhost download url directly on the server. You should see an empty response with a an X-Accel-Redirect header. Test when accessing through Nginx that the file is downloaded.

**Comment by keenangraham** [comment]
This looks okay but trying to trace through exactly what's happening. I'm assuming the `_make_location_absolute` is overriding this Webob response method: https://github.com/Pylons/webob/blob/6798afbce84d251b6aa1ba228acb150a2f03e143/src/webob/response.py#L1313

This looks like it is called from `_abs_headerlist`: https://github.com/Pylons/webob/blob/6798afbce84d251b6aa1ba228acb150a2f03e143/src/webob/response.py#L1325 if `location` is in a header key.

When I add a print statement and download a document from a demo I don't actually see this method getting triggered despite returning the `InternalRedirect` response:

```
def _make_location_absolute(environ, value):
        # Avoid changing Location header so Apache internally redirects.
        print('in _make_location_absolute', value)
        return value
```

I suppose the step I'm missing is that _instead_ of passing `X-Accel-Redirect` (which is an NGINX directive) you would pass `Location` which would trigger the Apache rewrite rule on /proxy? But you have to prevent the absolute URI from getting prepended so it stays internal? 

So to actually get this to do anything you would have to set `accel_redirect_header = X-Accel-Redirect` to `accel_redirect_header = Location`? In which case why call it accel_redirect?

What am I missing @lrowe?

**Comment by lrowe** [comment]
> So to actually get this to do anything you would have to set accel_redirect_header = X-Accel-Redirect to accel_redirect_header = Location?

Yes. I currently deploy encoded without nginx in front so would like the option to have the internal redirect happen in Apache.

As a bonus development instances can now run without nginx and will follow the redirect in Python.

> In which case why call it accel_redirect?

I couldn't think of a better name for what this does. The Location header behaviour is buried deep in the CGI spec. Perhaps `internal_redirect_header`?

**Comment by keenangraham** [comment]
By the way did you do something similar for files on your system? https://github.com/ENCODE-DCC/encoded/blob/ecfb135ba84ef5cad2a71638720d782cbfc4d14a/src/encoded/types/file.py#L789

**Comment by lrowe** [comment]
@keenangraham just made encode-dcc/encoded#3182 with the encoded change.

**Comment by keenangraham** [comment]
Rebased.

File: src/snovault/attachment.py [hunk 1]
```diff
@@ -37,6 +37,13 @@ def includeme(config):
     config.scan(__name__)
 
 
+class InternalRedirect(Response):
+    @staticmethod
+    def _make_location_absolute(environ, value):
+        # Avoid changing Location header so Apache internally redirects.
```
Comment (caseylitton) [inline src/snovault/attachment.py]: It looks like we're overriding the _make_location_absolute for downloads with 'get_blob_url' and 'accel_redirect_header'.  

The override is skipping this line of code in webob, https://github.com/Pylons/webob/blob/master/src/webob/response.py#L1314

What is the difference between returning the value regardless of the condition, 'if SCHEME_RE.searchv(value)' ?

Comment (lrowe) [reply src/snovault/attachment.py]:
```python
    @staticmethod
    def _make_location_absolute(environ, value):
        if SCHEME_RE.search(value):
            return value

        new_location = urlparse.urljoin(_request_uri(environ), value)
        return new_location
```
Looks like the `if SCHEME_RE.searchv(value)` is just a fast path for values which are already absolute urls (so would not be changed by urljoin.)

**Pull request merged**